MaxText/max_utils.py

```diff
@@ -351,3 +355,95 @@ def _cross_entropy_with_logits_bwd(
+  return optax.adamw(
+    learning_rate_schedule,
+    b1=config.adam_b1,
+    b2=config.adam_b2,
+    eps=config.adam_eps,
+    eps_root=config.adam_eps_root,
+    weight_decay=config.adam_weight_decay,
+  )
+
+def gen_shaped_input_data(model, tx, config, mesh):
+  example_rng = jax.random.PRNGKey(0)
+  shaped_rng = jax.ShapeDtypeStruct(example_rng.shape, example_rng.dtype)
+  abstract_state, state_mesh_annotations =  get_abstract_state(model, tx, config, example_rng, mesh)
+  shaped_batch = get_shaped_batch(config, np.size(mesh.device_ids))
+
+  input_args = (abstract_state, shaped_batch, shaped_rng)
+  input_kwargs = {}
+  return input_args, input_kwargs, state_mesh_annotations
+
+def get_shardings(mesh, state_mesh_annotations, config):
+  data_pspec = P(*config.data_sharding)
+  state_mesh_shardings = jax.tree_map(
+      lambda p: jax.sharding.NamedSharding(mesh, p), state_mesh_annotations)
+  data_sharding = jax.tree_map(
+      lambda p: jax.sharding.NamedSharding(mesh, p), data_pspec)
+  in_shardings = (state_mesh_shardings, data_sharding, None)
+  out_shardings = (state_mesh_shardings, None, None)
+  return in_shardings, out_shardings
+
+def get_abstract_state(model, tx, config, rng, mesh):
+  init_train_state_partial = functools.partial(init_train_state, model, tx,
+                                               config)
+  abstract_state = jax.eval_shape(init_train_state_partial, rng)
+  state_logical_annotations = nn.get_partition_spec(abstract_state)
+  unboxed_abstract_state = unbox_logicallypartioned_trainstate(abstract_state)
+
+  # Initialization
+  with mesh, nn_partitioning.axis_rules(config.logical_axis_rules):
+    state_mesh_annotations = nn.logical_to_mesh(state_logical_annotations)
+  return unboxed_abstract_state, state_mesh_annotations
+
+def get_shaped_batch(config, num_target_devices):
+  # Ahh this is bad. Cannot use local devices here since it is xaot - the target system may have a different
+  # count of local devices than the runner machine
+  
+  #batch_shape = (int(config.per_device_batch_size * num_target_devices), config.max_target_length)
+  # TODO(mattdavidow): confirm this should be global_batch_size_to_load and not global_batch_size_to_train_on
+  batch_shape = (config.global_batch_size_to_load, config.max_target_length)
+  print(f"{batch_shape=}", flush=True)
+  batch = {}
+  batch['inputs'] = jax.ShapeDtypeStruct(batch_shape, jnp.int32)
+  batch['inputs_position'] = jax.ShapeDtypeStruct(batch_shape, jnp.int32)
+  batch['inputs_segmentation'] = jax.ShapeDtypeStruct(batch_shape, jnp.int32)
+  batch['targets'] = jax.ShapeDtypeStruct(batch_shape, jnp.int32)
+  batch['targets_position'] = jax.ShapeDtypeStruct(batch_shape, jnp.int32)
+  batch['targets_segmentation'] = jax.ShapeDtypeStruct(batch_shape, jnp.int32)
+  print(f"{batch=}")
+  return batch
+
+
+
+# XAOT load
+def load_xaot(config, partial_train, state, num_target_devices):
```

Comment:
what is the safety vision on this? how can we ensure that it matches our jax/jaxlib/actual devices?

Reply:
load_xaot_train_step